asm-utils/src/data_parser.rs

```diff
@@ -14,76 +16,134 @@ pub enum DataValue {
-                            object_order.push(label);
-                            Default::default()
-                        })
-                        .extend(extract_data_value(dir.as_str(), args));
+                (".data", args) => {
+                    assert!(args.is_empty());
+                    is_in_data_section = true;
+                    data.append_section();
+                }
+                (".section", args) => {
+                    is_in_data_section = is_data_section(&args[0]);
+                    if is_in_data_section {
+                        data.append_section();
+                    }
                 }
                 (
-                    ".size",
-                    [Argument::Expression(Expression::Symbol(name)), Argument::Expression(Expression::Number(n))],
-                ) if Some(name.as_str()) == current_label => {
-                    let label = current_label.unwrap().to_string();
-                    objects
-                        .entry(current_label.unwrap().into())
-                        .and_modify(|entry| {
-                            let size: usize = entry.iter().map(|v| v.size()).sum();
-                            assert!(
-                                size as i64 == *n,
-                                "Invalid size for data object {name}: computed: {size} vs. specified: {n}"
-                            );
-                        })
-                        .or_insert_with(|| {
-                            object_order.push(label);
-                            assert!(*n == 0, "Nonzero size for object without elements: {name}");
-                            Default::default()
-                        });
+                    ".zero" | ".ascii" | ".asciz" | ".dword" | ".word" | ".half" | ".hword"
+                    | ".short" | ".byte",
+                    args,
+                ) => {
+                    if is_in_data_section {
+                        data.curr_entry()
+                            .extend(extract_data_value(dir.as_str(), args));
+                    } else {
+                        // This is most likely debug data.
+                    }
+                }
+                (".balign", [Argument::Expression(Expression::Number(byte_size))]) => {
+                    if is_in_data_section {
+                        data.curr_entry()
+                            .push(DataValue::Alignment(*byte_size as usize));
+                    }
+                }
+                (".p2align", [Argument::Expression(Expression::Number(pow_of_2))]) => {
+                    if is_in_data_section {
+                        data.curr_entry()
+                            .push(DataValue::Alignment((1 << pow_of_2) as usize));
+                    }
+                }
+                (".balign" | ".p2align", _) => {
+                    // TODO: implement the optional arguments of .balign and .p2align
+                    unimplemented!()
```

Comment:
What about `.type` and `.size` directives?

Reply:
Both are flags stored in the ELF and not validated by the assembler.

We used `.type` to identify the label type, i.e., if it is data or text. Now that role is performed by the `.section`.

`.size` was used to assert the size of the data pointed by label, if it happened to be the last label seen. But now `size()` needs the address, and we don't have it at this point.

I could still do the assert, but I would have to make the following assumptions that might not be true for every assembly code:
- there is actually exactly that many bytes between the label definition and the `.size`;
- there are no `.align` and `.p2align` between the label and the `.size`;
- the `.size` is actually inside the section of the label it refers too (not sure GNU `as` would accept differently).

Reply:
that's fine, just wanted to know what happened to them